OpenRA.Mods.Common/Activities/DonateExperience.cs

```diff
@@ -0,0 +1,53 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	class DonateExperience : Enter
+	{
+		readonly Actor target;
+		readonly int level;
+		readonly int playerExperience;
+
+		public DonateExperience(Actor self, Actor target, int level, int playerExperience)
+			: base(self, target, EnterBehaviour.Dispose)
+		{
+			this.target = target;
+			this.level = level;
+			this.playerExperience = playerExperience;
+		}
+
+		protected override void OnInside(Actor self)
+		{
+			if (target.IsDead)
+				return;
+
+			target.Trait<GainsExperience>().GiveLevels(level);
```

Comment:
Sorry, one last thing (now that I think of it). `target` is readonly and won't change therefore, so what about caching its `GainsExperience` trait in the constructor? We could then remove the `.Trait<>` calls from here and from `Tick`.

I made an example for that: https://github.com/abcdefg30/OpenRA/commit/837653862b4b1c7a04e861a20e18d384db1ab8e2
(The example also already caches the trait in `ResolveOrder` from `DeliversExperience` to save 2 more trait lookups.)

You should at least remove the unused `OpenRA.Mods.Common.Effects` and `OpenRA.Traits` usings from this file, I think.